### ISSUE TITLE
HTTPS auth doesn't work

### ISSUE THREAD

**Issue by chrisburrus** [concourse/git-resource#65]
Git resources using https auth don't seem to function at all.

Using this setup:

```
resources:
- name: reponame
  type: git
  source:
    uri: https://github.com/companyname/reponame.git
    branch: master
    username: githubusername
    password: githubpassword
```

The check script fails with the following error message:

```
resource script '/opt/resource/check []' failed: exit status 128

stderr:
Cloning into '/tmp/git-resource-repo-cache'...
fatal: could not read Username for 'https://github.com': No such device or address
```

HTTPS auth is important because you can't authenticate submodules using private keys. Thanks!

**Comment by chrisburrus** [comment]
As a note, it works if you pass the credentials in explicitly in the uri:

`uri: https://username:password@github.com/companyname/reponame.git`

But then the submodules fail to load, as the credentials attempt to load via ssh:

```
Submodule 'src/subrepo' (git@github.com:company/subrepo.git) registered for path 'src/subrepo
Cloning into 'src/subrepo'...
Host key verification failed.
fatal: Could not read from remote repository.
```

**Comment by vito** [comment]
@MatthiasWinzeler any thoughts on this? we've never used it ourself. does it work for you?

**Comment by MatthiasWinzeler** [comment]
the mechanism definitely works - we have tons of git resources that successfully fetch git repos from our internal gitlab using http auth.

I quickly checked, it also works fine with github. 

@chrisburrus Can you double check your password? I receive your exact error when I try with a wrong password. Makes sense, since git tries to ask the user for the correct password and can't allocate a tty (in the container).

**Comment by chrisburrus** [comment]
@MatthiasWinzeler Based on that information, I tried a fresh account, and it worked just fine. I eventually realized what the original problem was - the problem was that 2FA was set up on the account. Is there a way to use an account with 2FA and https authentication?

But past that you are correct, it otherwise works just fine.

**Comment by vito** [comment]
That makes sense then. There's no way to automate 2FA (kinda the whole point), so you should probably be using private keys instead.

**Comment by MatthiasWinzeler** [comment]
@chrisburrus : When using 2FA, you can use [Github's personal access tokens](https://help.github.com/articles/which-remote-url-should-i-use/#cloning-with-https-urls-recommended) instead of the password. Just tested it and it works fine.

**Comment by vito** [comment]
Ah, neat. Good to know!

**Comment by mxplusb** [comment]
Wanted to update on this issue. I have the same issue, but I don't have 2FA enabled.

This doesn't work and fails with the same error:

```yaml
- name: repo
  type: git
  source:
    uri: https://git.private.org/path/repo.git
    branch: master
    username: ((git.username)) # interpolated to <user>
    password: ((git.password)) # interpolated to <password>
```

But this does (sporadically - might just be my setup):

```yaml
- name: repo
  type: git
  source:
    uri: https://((git.username)):((git.password))@git.private.org/path/repo.git # interpolated properly
    branch: master
```

The private repository is with Gitlab, not Github, and since 2FA/SSH isn't enabled, cloning with a private key or personal token isn't allowed. The problem is if there is an error in the git resource, it will dump the credentials into the stderr of the resource:

```
stderr:
Cloning into '/tmp/git-resource-repo-cache'...
fatal: could not read Username for 'https://((git.username)):((git.password))@git.private.org': No such device or address
```

If you look at the [git-credential-store](https://git-scm.com/docs/git-credential-store) documentation, instead of populating `~/.netrc`, you should be able to populate `~/.git-credentials` with `https://user:pass@domain.com` to solve the problem.

**Comment by MatthiasWinzeler** [comment]
@mxplusb This is strange - we use it with Gitlab as well and it works fine. Do you suggest changing it to `git-credential-store`? Did you test that this works for your setup?

**Comment by mxplusb** [comment]
@MatthiasWinzeler I did try, but I wasn't able to preseed credentials. See [here](https://github.com/concourse/git-resource/compare/master...mxplusb:master#diff-1623687d4df544193e2c30926ea689bcR144). I'm not sure if it's the way `git-credential-store` operates or something else, because it never picked up the credentials.

**Comment by tchaloupka** [comment]
Same problem as @mxplusb described here. But with repository hosted on visualstudio.com.
When user/password is in url, it works (but not with submodules) when set as `username` and `password` parameters it ends up with
```
fatal: could not read Username for 'https://xxx.visualstudio.com': No such device or address
```

**Comment by MatthiasWinzeler** [comment]
@tchaloupka @mxplusb If you hijack the container and run the clone manually, does it work? 
I suspect `No such device or address` comes from a password prompt - so I assume this happens in case auth is unsuccessful (being due to 2FA enabled or wrong credentials) and it falls back to asking for credentials.

**Comment by mxplusb** [comment]
@MatthiasWinzeler, no, I can't clone it locally. I wonder if you need to pass `DEBIAN_FRONTEND=noninteractive`? That's the only thing I can think of that would cause it.

**Comment by MatthiasWinzeler** [comment]
@mxplusb If you set this env var, does it work (manually)?

**Comment by larssb** [comment]
I have the exact same issue. See [Semver issue with git](https://github.com/concourse/semver-resource/issues/80).

Is this solved. There has been somewhat recent communication here. If so what is the solution?

**Comment by larssb** [comment]
Update. I closed [#80](https://github.com/concourse/semver-resource/issues/80) in the semver-resource. I found a culprit and will report a bug report on it.

**Comment by nickmly** [comment]
> Same problem as @mxplusb described here. But with repository hosted on visualstudio.com.
> When user/password is in url, it works (but not with submodules) when set as `username` and `password` parameters it ends up with
> 
> ```
> fatal: could not read Username for 'https://xxx.visualstudio.com': No such device or address
> ```
> No 2FA used here, just generated access token

Did you have any luck with this? We are having the exact same issue with a visualstudio.com repo, however it doesn't work with the user/password in the url either
